### PR TITLE
[NPUW] Add Slice before last MatMul

### PR DESCRIPTION
**Pull request by smirnov-alexey**
Based on https://github.com/openvinotoolkit/openvino.genai/pull/814

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 1]
```diff
@@ -1296,6 +1285,147 @@ CompressDictMatMulf32::CompressDictMatMulf32(Context::Ref ctx) {
...
+            return true;  // root was changed
+        }
+        return false;  // root hasn't changed
+    };
+    register_matcher(std::make_shared<opp::Matcher>(res, "SliceLastMatmul"), std::move(callback));
+}
+
+SliceLastMatmulAdd::SliceLastMatmulAdd() {
+    auto matmul = opp::wrap_type<ov::op::v0::MatMul>({opp::any_input(), opp::any_input()});
+    auto add = opp::wrap_type<ov::op::v1::Add>({matmul, opp::any_input()});
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1324]: Not sure if all those additional patterns could be simplified with `optional` nodes

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 2]
```diff
@@ -1296,6 +1285,151 @@ CompressDictMatMulf32::CompressDictMatMulf32(Context::Ref ctx) {
...
+    auto res = opp::wrap_type<ov::op::v0::Result>({add});
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1326]: Not sure if I've seen cases like this. @TolyaTalamanov did you?

@smirnov-alexey  what pattern worked for you most of the time?

I think this pattern we can drop

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 3]
```diff
@@ -1296,6 +1285,151 @@ CompressDictMatMulf32::CompressDictMatMulf32(Context::Ref ctx) {
...
+        }
+        return false;  // root hasn't changed
+    };
+    register_matcher(std::make_shared<opp::Matcher>(res, "SliceLastMatmulAdd"), std::move(callback));
+}
+
+SliceLastMatmulTranspose::SliceLastMatmulTranspose() {
+    auto matmul = opp::wrap_type<ov::op::v0::MatMul>({opp::any_input(), opp::any_input()});
+    auto add = opp::wrap_type<ov::op::v1::Transpose>({matmul, opp::any_input()});
+    auto res = opp::wrap_type<ov::op::v0::Result>({matmul});
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1362]: Also not sure about this one. My main concern is that we can alter more matmuls than we actually need..

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 4]
```diff
@@ -1296,6 +1285,151 @@ CompressDictMatMulf32::CompressDictMatMulf32(Context::Ref ctx) {
...
+                std::make_shared<ov::op::v8::Slice>(matched_out_matmul.get_node()->input_value(0), start, stop, step);
+
+            matched_out_matmul.get_node()->input(0).replace_source_output(slice);
+
+            return true;  // root was changed
+        }
+        return false;  // root hasn't changed
+    };
+    register_matcher(std::make_shared<opp::Matcher>(res, "SliceLastMatmulMultiply"), std::move(callback));
+}
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1431]: Same concern here. Not sure which topologies does it serve.

File: src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp [hunk 5]
```diff
@@ -147,6 +147,14 @@ ov::npuw::CompiledModel::CompiledModel(const std::shared_ptr<ov::Model>& model,
         rewr.run_on_model(model);
     }
 
+    // Add Slice before last MatMul for the prefill model
+    ov::pass::GraphRewrite rewr;
+    rewr.add_matcher<ov::npuw::patterns::opt::SliceLastMatmul>();
+    rewr.add_matcher<ov::npuw::patterns::opt::SliceLastMatmulAdd>();
+    rewr.add_matcher<ov::npuw::patterns::opt::SliceLastMatmulTranspose>();
+    rewr.add_matcher<ov::npuw::patterns::opt::SliceLastMatmulMultiply>();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp:155]: I'd drop these three

File: src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp [hunk 6]
```diff
@@ -147,6 +147,14 @@ ov::npuw::CompiledModel::CompiledModel(const std::shared_ptr<ov::Model>& model,
         rewr.run_on_model(model);
     }
 
+    // Add Slice before last MatMul for the prefill model
+    ov::pass::GraphRewrite rewr;
+    rewr.add_matcher<ov::npuw::patterns::opt::SliceLastMatmul>();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp:152]: Let's make this under a config too. We're breaking the model's API convention in fact, user gave us a model with `[N,K]` output, but the compiled model he gets instead would be `[1,K]`.

Let it be `NPUW_SLICE_OUT`, let it be NO by default, let @TolyaTalamanov and our internal configs pass it by default.

**Review by dmatveev — APPROVED** [review]
Fantastic work! @slyalin can you please comment on my concerns?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 2 above)
Comment (slyalin) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1326]: Summoning @olpipi who showed that this and later mentioned was required. @olpipi could you give a list of topologies?

Comment (slyalin) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1326]: phi2: https://huggingface.co/OpenVINO/phi-2-fp16-ov/raw/main/openvino_model.xml has `Add` after `MatMul`.

File: src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp (hunk 6 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 4 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1431]: As discussed, keeping for now

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 3 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1362]: As discussed, keeping for now

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 2 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1326]: I only saw the first one, but as discussed let's keep all 3 for now (enabled via property)

**Review by dmatveev — APPROVED** [review]

**Review by dmatveev — APPROVED** [review]

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 2 above)
Comment (olpipi) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp:1326]: I found the next patterns and applied them in genai repo:

chatglm2-6b
chatglm3-6b
Matmul -> Transpose -> Result

codegen2-1b
codegen2-3_7b
codegen2-7b
gpt-j-6b
phi-2
Matmul -> Add -> Result

gemma-2-2b
gemma-2-9b
MatMul -> Divide -> Tanh -> Multiply -> Result

**Pull request merged**